code/citadel/hypospraymkii.dm

```diff
@@ -0,0 +1,238 @@
+#define HYPO_SPRAY 0
+#define HYPO_INJECT 1
+
+//A vial-loaded hypospray. Cartridge-based!
+/obj/item/reagent_containers/hypospray/mkii
+	name = "hypospray mk.II"
+	icon = 'icons/obj/citadel/hypospray.dmi'
+	icon_state = "hypo2"
+	var/list/allowed_containers = list(/obj/item/reagent_containers/glass/bottle/vial)
+	desc = "A new development from DeForest Medical, this new hypospray takes 30-unit vials as the drug supply for easy swapping."
+	volume = 0
+	amount_per_transfer_from_this = 5
+	possible_transfer_amounts = list(5,10,15)
+	var/mode = HYPO_INJECT
+	var/obj/item/reagent_containers/glass/bottle/vial
+	var/loaded_vial = /obj/item/reagent_containers/glass/bottle/vial
+	var/spawnwithvial = TRUE
+	var/start_vial = null
+	var/list/vials = list()
+	var/list/inject_sounds = list('sound/items/hypospray.ogg','sound/items/hypospray2.ogg')
+
+/obj/item/reagent_containers/hypospray/mkii/CMO
+	name = "hypospray mk.II deluxe"
+	allowed_containers = list(/obj/item/reagent_containers/glass/bottle/vial, /obj/item/reagent_containers/glass/bottle/vial/large)
+	icon_state = "cmo2"
+	ignore_flags = 1
+	desc = "The Chief Medical Officer's hypospray is identically functional to the base model, excepting that it can take larger vials in addition to regular sized. It is also able to penetrate harder materials and deliver more reagents per spray."
+	resistance_flags = INDESTRUCTIBLE | LAVA_PROOF | FIRE_PROOF | ACID_PROOF
+	loaded_vial = /obj/item/reagent_containers/glass/bottle/vial/large/preloaded/CMO
+	possible_transfer_amounts = list(5,10,15,30,60) //cmo hypo should be able to dump lots into it
+
+/obj/item/reagent_containers/hypospray/mkii/Initialize()
+	. = ..()
+	if(!spawnwithvial)
+		update_icon()
+		return
+	if (!start_vial)
+		start_vial = new loaded_vial(src)
+	update_icon()
+
+/obj/item/reagent_containers/hypospray/mkii/update_icon()
+	..()
+	icon_state = "[initial(icon_state)][loaded_vial ? "" : "-e"]"
+	if(ismob(loc))
+		var/mob/M = loc
+		M.update_inv_hands()
+	return
+
+/obj/item/reagent_containers/hypospray/mkii/examine(mob/user)
+	. = ..()
+	to_chat(user, "[src] is set to [mode ? "Inject" : "Spray"] contents on application.")
+
+/obj/item/reagent_containers/hypospray/mkii/proc/unload_hypo(obj/item/reagent_containers/glass/bottle/vial/W, mob/user as mob)
+	if(!W)
+		to_chat(user, "This doesn't have a vial!")
+		return
+	testing("passed hypo check in unload_hypo")
+	vial = W
+	testing("vial = W")
+	reagents.trans_to(vial, volume)
+	volume = 0
+	testing("attempting forcemove")
+	vial.forceMove(drop_location())
+	testing("forceMove valid, testing put_in_hands")
+	user.put_in_hands(vial)
+	testing("put_in_hands passed")
+	contents -= vial
```

Comment:
Try using transferitemtoloc to move vials in and out of hyposprays instead of this